src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -0,0 +1,426 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import net.snowflake.ingest.streaming.OpenChannelRequest;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestChannel;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClientFactory;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is per task configuration. A task can be assigned multiple partitions. Major methods are
+ * startTask, insert, getOffset and close methods.
+ *
+ * <p>StartTask: Called when partitions are assigned. Responsible for generating the POJOs.
+ *
+ * <p>Insert and getOffset are called when {@link
+ * com.snowflake.kafka.connector.SnowflakeSinkTask#put(Collection)} and {@link
+ * com.snowflake.kafka.connector.SnowflakeSinkTask#preCommit(Map)} APIs are called.
+ *
+ * <p>This implementation of SinkService uses Streaming Snowpipe (Streaming Ingestion)
+ *
+ * <p>Hence this initializes the channel, opens, closes. The StreamingIngestChannel resides inside
+ * {@link TopicPartitionChannel} which is per partition.
+ */
+public class SnowflakeSinkServiceV2 implements SnowflakeSinkService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SnowflakeSinkServiceV2.class);
+
+  private static String STREAMING_CLIENT_PREFIX_NAME = "KC_CLIENT_";
+
+  // Assume next three values are a threshold after which we will call insertRows API
+  // Set in config (Time based flush) in seconds
+  private long flushTime;
+  // Set in config (buffer size based flush) in bytes
+  private long fileSize;
+
+  // Set in config (Threshold before we call insertRows API) corresponds to # of
+  // records in kafka
+  private long recordNum;
+
+  // Used to connect to Snowflake
+  private final SnowflakeConnectionService conn;
+
+  private final RecordService recordService;
+  private final SnowflakeTelemetryService telemetryService;
+  private Map<String, String> topic2TableMap;
+
+  // Behavior to be set at the start of connector start. (For tombstone records)
+  private SnowflakeSinkConnectorConfig.BehaviorOnNullValues behaviorOnNullValues;
+
+  // default is true unless the configuration provided is false;
+  // If this is true, we will enable Mbean for required classes and emit JMX metrics for monitoring
+  private boolean enableCustomJMXMonitoring = SnowflakeSinkConnectorConfig.JMX_OPT_DEFAULT;
+
+  // default is at_least_once semantic (To begin with)
+  // TODO: SNOW-526435
+  private SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee ingestionDeliveryGuarantee =
+      SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.AT_LEAST_ONCE;
```

Comment:
this won't apply here, right?

Reply:
I am hoping it doesnt apply here, but I havent much thought about this config for Streaming Ingestion. 

I know internally we do have rowSequencer which will help us get to EXACTLY_ONCE.

Also, this member (ingestionDeliveryGuarantee) isnt used anywhere in this file. So we are safe for now and I have logged the JIRA to tackle this separately. 

Potentially we will not support this config if streaming ingestion is used.

Reply:
I'd say let's do `EXACTLY_ONCE` out the gate. It'll be weird otherwise to go to at least once using snowpipe (current), exactly once using snowpipe (being rolled out), at least once using streaming ingest (this pr), exactly once using streaming ingest. Actually, that would be a regression in behavior and we should definitely avoid that.

Reply:
That is true.. Will keep this in mind and tackle the JIRA SNOW-526435 in separate PR.

Reply:
We do have logic in place based on what we implemented for snowpipe and also we have similar APIs. so not too tough to do `exactly_once` from initial look.